Model instance v2

rudacs opened the pull request:
# Description

Some adjustments to the implementation being made to ModelInstanceCollection

ggetz commented:
Thank you for the fixes here @rudacs! I can confirm we have a CLA on file for you and Geoex Tecnologias.

@lukemckinstry Could you please review and merge into your branch?

lukemckinstry commented:
Thanks @rudacs for this PR. Two items 

 - I needed to rebase `model-instance-v2` to fix conflicts in `main`. Looks like this impacted the git history in this branch here, so I think you will want to rebase onto the latest version of `model-instance-v2` as well. I suggest doing a interactive rebase (`git rebase -i model-instance-v2`) and dropping all but the original 6 commits in your branch.
 - To make sure I am understanding you correctly, how are you using and testing the pick logic you added? Can you provide some test instructions for that part?

rudacs commented:
> * To make sure I am understanding you correctly, how are you using and testing the pick logic you added? Can you provide some test instructions for that part?

:)

The idea was that when I pick I would know which instance I was clicking on.

```
const picked = viewer.scene.pick(movement.position);
const modelInstance = picked.primitive.instances.get(picked.instanceId);
```

With the pickId in the ModelInstance, I can:

```
const edge = Cesium.PostProcessStageLibrary.createEdgeDetectionStage();
edge.uniforms.color = Cesium.Color.RED;
edge.uniforms.length = 0.05;
edge.selected = [modelInstance];

viewer.scene.postProcessStages.add(Cesium.PostProcessStageLibrary.createSilhouetteStage([edge]));
```

I also exposed the transform so I could, for example, change its position.

````
modelInstance.transform = newModelMatrix;
````

I also adjusted the remove because as I don't know what the index is and the add returns the ModelInstance, I changed the remove to pass the instance.

```
const modelInstance = collection.add(...);
collection.remove(modelInstance);
```

lukemckinstry commented on packages/engine/Source/Scene/Model/PickingPipelineStage.js:
```diff
@@ -38,11 +38,11 @@ PickingPipelineStage.process = function (
   if (renderResources.hasPropertyTable) {
     processPickTexture(renderResources, primitive, instances, context);
-  } else if (defined(instances)) {
+  } else if (defined(instances) && instances > 0) {
```
We actually have slightly different code paths for instancing from glTF and i3dm versus instancing from the API which we have started to call runtime instancing. 
I believe `runtimeNode.node.instances` supports the glTF andf i3dm case. And it looks like  `processPickTexture` is expecting `runtimeNode.node,instances`.
So I think adding
```
const instanceCount =  renderResources.instanceCount;
```
and then modifying the lower line to 
```
} else if (defined(instances) || (defined(instanceCount) && instanceCount > 0)) {
```

would cover both cases.

rudacs replied:
Adjusted.
It would actually fail for another case.
